src/esp/core/Buffer.h

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+#include "esp/core/esp.h"
+
+namespace esp {
+namespace core {
+
+// Enumeration of data
+enum class DataType {
+  DT_NONE = 0,
+  DT_INT8 = 1,
+  DT_UINT8 = 2,
+  DT_INT16 = 3,
+  DT_UINT16 = 4,
+  DT_INT32 = 5,
+  DT_UINT32 = 6,
+  DT_INT64 = 7,
+  DT_UINT64 = 8,
+  DT_FLOAT = 9,
+  DT_DOUBLE = 10,
+};
+
+class Buffer {
```

Comment:
Any reason why this takes a datatype enum instead of being templated  i.e if you want a double buffer, just get a `Buffer<double>`?

Reply:
Never mind, this makes sense with the way it's being used.

Reply:
> Maybe it makes sense to take the current thing we call the simulator and rename it `HabitatSimBackend`? Then we can have this the `c++` front-end and the existing thing be the `python` front end?

Since the js work is new and exploratory, the approach we are taking with the js port  is to extend the c++ port first and then rationalize the python port once we have a good understand of the architecture. That way we avoid churn and destabilization of the python port during the exploratory phase. The downside is that you don't get the benefits of an integrated code base (testing, etc). Does that sound reasonable?

Reply:
> Any reason why this takes a datatype enum instead of being templated i.e if you want a double buffer, just get a `Buffer<double>`?

We don't know the type at compile time.